src/main/scala/scorex/transaction/state/database/blockchain/StoredState.scala

```diff
@@ -1,37 +1,223 @@
-                  val leaseExtendedState: LeaseExtendedState,
-                  val assetsExtension: AssetsExtendedState,
-                  val incrementingTimestampValidator: IncrementingTimestampValidator,
-                  val validators: Seq[Validator],
+class StoredState(private val storage: StateStorageI with AssetsExtendedStateStorageI with OrderMatchStorageI
+  with LeaseExtendedStateStorageI with AliasExtendedStorageI,
                   settings: ChainParameters) extends State with ScorexLogging {
 
-  override def included(id: Array[Byte]): Option[Int] = storage.included(id, None)
+  def applyAssetIssueReissueBurnTransaction(height: Int)(tx: Transaction): Unit = tx match {
+    case tx: AssetIssuance =>
+      addAsset(tx.assetId, height, tx.id, tx.quantity, tx.reissuable)
+    case tx: BurnTransaction =>
+      burnAsset(tx.assetId, height, tx.id, -tx.amount)
+    case _ =>
+  }
+
+  def addAsset(assetId: AssetId, height: Int, transactionId: Array[Byte], quantity: Long, reissuable: Boolean): Unit = {
+    val asset = Base58.encode(assetId)
+    val transaction = Base58.encode(transactionId)
+    val assetAtHeight = s"$asset@$height"
+    val assetAtTransaction = s"$asset@$transaction"
+
+    if (storage.getHeights(Base58.encode(assetId)).isEmpty ||
+      (reissuable && isReissuable(assetId)) ||
+      !reissuable) {
+      storage.setReissuable(assetAtTransaction, reissuable)
+    } else {
+      throw new RuntimeException("Asset is not reissuable")
+    }
+    storage.addHeight(asset, height)
+    storage.addTransaction(assetAtHeight, transaction)
+    storage.setQuantity(assetAtTransaction, quantity)
+    storage.setReissuable(assetAtTransaction, reissuable)
+  }
+
+  def burnAsset(assetId: AssetId, height: Int, transactionId: Array[Byte], quantity: Long): Unit = {
+    require(quantity <= 0, "Quantity of burned asset should be negative")
+
+    val asset = Base58.encode(assetId)
+    val transaction = Base58.encode(transactionId)
+    val assetAtHeight = s"$asset@$height"
+    val assetAtTransaction = s"$asset@$transaction"
+
+    storage.addHeight(asset, height)
+    storage.addTransaction(assetAtHeight, transaction)
+    storage.setQuantity(assetAtTransaction, quantity)
+  }
+
+  def assetRollbackTo(assetId: AssetId, height: Int): Unit = {
+    val asset = Base58.encode(assetId)
+
+    val heights = storage.getHeights(asset)
+    val heightsToRemove = heights.filter(h => h > height)
+    storage.setHeight(asset, heights -- heightsToRemove)
+
+    val transactionsToRemove: Seq[String] = heightsToRemove.foldLeft(Seq.empty[String]) { (result, h) =>
+      result ++ storage.getTransactions(s"$asset@$h")
+    }
+
+    val keysToRemove = transactionsToRemove.map(t => s"$asset@$t")
+
+    keysToRemove.foreach { key =>
+      storage.removeKey(key)
+    }
+  }
+
+  def isReissuable(assetId: AssetId): Boolean = {
+    val asset = Base58.encode(assetId)
+    val heights = storage.getHeights(asset)
+
+    heights.lastOption match {
```

Comment:
This is not a new code. A lot of methods like `getAssetQuantity`, `isReissuable`, `getAssetName` etc will hopefully be merged and completely rewritten to  a single `getAssetInfo`, hence minimal changes now